File: 프로그래머스/키패드_누르기.java
package 프로그래머스;

public class 키패드_누르기 {
    public static void main(String[] args){
        int[] numbers = {7,0,8,2,8,3,1,5,7,6,2};
        String hand = "left";
        System.out.println(solution(numbers,hand));
    }
    public static String solution(int[] numbers, String hand) {
        StringBuilder answer = new StringBuilder();
        int leftHand = 10;
        int rightHand = 12;
        for (int i = 0; i < numbers.length; i++){
            switch (numbers[i]){
                case 1,4,7:
                    answer.append('L');
                    leftHand = numbers[i];
                    break;

                case 3,6,9:
                    answer.append('R');
                    rightHand = numbers[i];
                    break;

                default:
                    if (numbers[i] == 0){
                        numbers[i] = 11;
                    }

                    int leftDiff = Math.abs(numbers[i] - leftHand)/3 + Math.abs(numbers[i] - leftHand)%3;
                    int rightDiff = Math.abs(numbers[i] - rightHand)/3 + Math.abs(numbers[i] - rightHand)%3;

                    if (rightDiff < leftDiff){
                        answer.append('R');
                        rightHand = numbers[i];

                    } else if (rightDiff == leftDiff) {
                        if (hand.equals("left")){
                            answer.append('L');
                            leftHand = numbers[i];
                        }else{
                            answer.append('R');
                            rightHand = numbers[i];
                        }

                    }else{
                        answer.append('L');
                        leftHand = numbers[i];
                    }
                    break;
            }
        }
        return answer.toString();
    }
}

File: 프로그래머스/키패드_누르기.py
def solution(numbers, hand):
    answer = ''
    leftHand = 10
    rightHand = 12
    for n in numbers:
        if n == 1 or n == 4 or n == 7:
            answer += "L"
            leftHand = n
        elif n == 3 or n == 6 or n == 9:
            answer += "R"
            rightHand = n
        else:
            if n == 0:
                n = 11

            leftDiff = abs(n - leftHand)//3 + abs(n - leftHand)%3
            rightDiff = abs(n - rightHand)//3 + abs(n - rightHand)%3

            if leftDiff > rightDiff:
                answer += "R"
                rightHand = n
            elif leftDiff == rightDiff:
                if hand == "right":
                    rightHand = n
                    answer += "R"
                else:
                    answer += "L"
                    leftHand = n
            else:
                answer += "L"
                leftHand = n
    return answer

print(solution([7, 0, 8, 2, 8, 3, 1, 5, 7, 6, 2], "left"))